rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,131 @@
+reusable kernels to be dispatched preferentially over the standard
+performance-optimized kernel ordering.
+
+> **Note:** Reusable kernels are defined as kernels which have internally
+opted-into the kernel cache API, and support a notable subset of primitive
+configurations with a single compilation. Reusable kernels may depend on
+problem shape, but only in a limited way. For example, a reusable kernel may
+impose a divisibility requirement (e.g. input channels must be a multiple of
+16), but the specific shape cannot be compiled directly into the kernel.
+
+## Proposal
+
+The proposal has four potential options: using DNNL_RUNTIME_DIM_VAL, a build
+flag, an environment variable, or a primitive attribute. In any of the options,
+the default behaviour will be the current one: kernels are ordered according to
+internal heuristics, roughly in order of decreasing likelihood for optimal
+performance for the given problem.
+
+If the reusable kernel option is selected, kernels designed to be reusable will
+be selected first. In the case that no reusable kernels are compatible with the
+problem, kernel selection will continue with non-reusable kernels.
+
+Option 4 (primitive attribute) is the preferred option by frameworks, as they
+don't know which dimensions should be considered dynamic at runtime. It
+provides a balance between user control, flexibility, and ease of integration.
+Compared to the build flag it's more flexible, and compared to the environment
+variable it's less opaque.
+
+### Option 1: Using `DNNL_RUNTIME_DIM_VAL`
+
+With this option, the user will select one or more dimensions to have the
+`DNNL_RUNTIME_DIM_VAL` special value during primitive creation. oneDNN will
+then dispatch to the first implementation that supports the given set of
+runtime dimensions. One downside to this approach is that the implementation
+must either compile a kernel that's general enough to support any possible
+value for the runtime dimension (which limits optimization potential) or it
+must identify and compile all specialized kernels which support the given set
+of runtime dimensions (which would increase primitive creation time).
+
+In addition to existing limitations on primitives that support runtime
+dimensions, feedback from frameworks has been that which dimensions should be
+specified as `DNNL_RUNTIME_DIM_VAL` is unknown at primitive creation time.
+Between these two factors, this option is unlikely to lead to a satisfactory
+solution, at least in the short term.
+
+<u>Pros:</u>
+- The most fine-grained approach
+- Ties into existing mechanisms
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- Several primitives have no implementations that support runtime dims -
+cannot be adopted until such implementations have been developed
+- Support for all possible runtime dim configurations is difficult
+- Requires users to know which dimensions will have dynamic shapes ahead of
+time
```

Comment:
> From FWK perspective, it is never known which shapes will effectively be dynamic since this is an application property:

This is a good point. I would check with the frameworks folks whether the hint option can really work for them (e.g. hypothetically then can delegate setting this hint to the end user). If the hint doesn't look good from their side, I'm afraid we don't have any realistic alternatives at this point besides gradually closing the gap between JIT/reusable path and making the reusable path default starting from some architecture in the future.